staging/src/k8s.io/apiserver/pkg/util/flowcontrol/dropped_requests_tracker.go

```diff
@@ -0,0 +1,184 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flowcontrol
+
+import (
+	"sync"
+
+	"k8s.io/utils/clock"
+)
+
+const (
+	// maxHistory represents what is the maximum history of dropped
+	// requests stored per priority level.
+	maxHistory = 30
+)
+
+var (
+	// retryAfterSteps are the possible values for retryAfter.
+	retryAfterSteps = []int64{1, 2, 3, 5, 8, 13}
+)
+
+// DroppedRequestsTracker is an interface that allows tracking
+// a history od dropped requests in the system for the purpose
+// of adjusting RetryAfter header to avoid system overload.
+type DroppedRequestsTracker interface {
+	// RecordDroppedRequest records a request that was just
+	// dropped from processing.
+	RecordDroppedRequest(plName string)
+
+	// GetRetryAfter returns the current suggested value of
+	// RetryAfter value.
+	GetRetryAfter(plName string) int64
+}
+
+// unixStat keeps a statistic how many requests were dropped within
+// a single second.
+type unixStat struct {
+	unixSecond int64
+	requests   int64
+}
+
+type droppedRequestsStats struct {
+	history []unixStat
+
+	currentRetryAfterStep int
+	retryAfterUpdateUnix  int64
```

Comment:
```suggestion
        // retryAfterUpdateUnix is the time when currentRetryAfterStep was last updated, in seconds since Unix epoch.
	retryAfterUpdateUnix  int64
```